corehq/tests/optimizer.py

```diff
@@ -0,0 +1,152 @@
+
+    class MyTestCase(TestCase):
+        dependent_apps = ['myapp.dependentapp1', 'myapp.dependentapp2']
+
+    This will override the app-level setting. Like apps, tests are always assumed to be dependent
+    on the apps they live in.
+
+    Updating:
+
+    The easiest way to determine an app/test's dependencies is to mark them empty and then run the tests.
+    The tests will fail on various django/couch access. You can then iteratively add those apps to the
+    list of dependencies until tests pass again.
+    """
+
+    def run_tests(self, test_labels, extra_tests=None, **kwargs):
+        test_labels = test_labels or self.get_test_labels()
+        with optimize_apps_for_test_labels(test_labels):
+            return super(OptimizedTestRunnerMixin, self).run_tests(test_labels, extra_tests, **kwargs)
+
+
+class AppAndTestMap(object):
+
+    def __init__(self):
+        self._dependencies = get_app_test_db_dependencies()
+        # these are permanently needed apps
+        self.required_apps = set([
+            'django.contrib.auth',
+            'django.contrib.contenttypes',
+        ])
+        self.apps = set([])
+        self.tests = []  # contains tuples of app_labels and test classes
+
+    def add_test(self, app_label, test):
+        self.tests.append((app_label, test))
+
+    def add_app(self, app_label):
+        self.apps.add(app_label)
+
+    def get_needed_installed_apps(self):
+        try:
+            needed_apps = copy(self.required_apps)
+            for app in self.apps:
+                needed_apps.update(self.get_app_dependencies(app))
+
+            for app, test in self.tests:
+                needed_apps.update(self.get_test_dependencies(app, test))
+
+            return needed_apps
+        except DependenciesNotFound:
+            # any time we can't detect dependencies fall back to including all apps
+            return settings.INSTALLED_APPS
+
+    def get_app_dependencies(self, app):
+        if app not in self._dependencies:
+            raise DependenciesNotFound()
+        return self._dependencies[app]
+
+    def get_test_dependencies(self, app, test):
+        dependencies = set([apps.get_app_config(app).name])
+
+        def _extract_tests(test_suite_or_test):
+            if isinstance(test_suite_or_test, unittest.TestCase):
+                return [test_suite_or_test]
+            elif isinstance(test_suite_or_test, unittest.TestSuite):
+                return test._tests
+            else:
+                raise DependenciesNotFound()
+
+        for test in _extract_tests(test):
+            if hasattr(test, 'dependent_apps'):
+                dependencies.update(test.dependent_apps)
+            else:
+                # if any of the tests don't have explicit dependencies defined
+                # immediately short-circuit to using all app dependencies for the app
+                return self.get_app_dependencies(app)
+        return dependencies
+
+
+class optimize_apps_for_test_labels(object):
```

Comment:
neat - I'd never seen that before. https://github.com/dimagi/commcare-hq/pull/9242